[Core] Report SDK resolution errors

mrward opened the pull request:
If an SDK cannot be resolved using the MSBuild SDK resolvers then
an exception is now thrown so the error is reported in the IDE.
Previously the error would be logged in the IDE log and the error
displayed in the IDE would usually be a confusing 'Invalid
configuration mapping'. This also makes the behaviour more consistent
with Visual Studio on Windows.

Before:

No load error. Solution pad shows:

<img width="297" alt="CurrentBehaviour" src="https://user-images.githubusercontent.com/372361/59422983-24643280-8dc9-11e9-84ae-8183a2889e38.png">

After:

Load error dialog on opening solution.

<img width="606" alt="SdkNotFoundLoadError" src="https://user-images.githubusercontent.com/372361/59592050-2985f200-90e7-11e9-9deb-d48d615cca2d.png">

Solution pad error tooltip.

<img width="344" alt="SolutionPadSdkNotFoundError" src="https://user-images.githubusercontent.com/372361/59592086-3f93b280-90e7-11e9-9850-c8e7c441c275.png">

Multiple project load errors.

<img width="566" alt="MultipleSdkNotFoundErrors" src="https://user-images.githubusercontent.com/372361/59592121-4cb0a180-90e7-11e9-93f9-8ab5745bb676.png">


Note that the error text here is generated by the SDK resolvers not by VS Mac.

VS Win:

Does not show error information on loading the solution.

<img width="415" alt="VSWinProjectLoad" src="https://user-images.githubusercontent.com/372361/59423270-b835fe80-8dc9-11e9-81de-353494459fc0.png">

On reloading the project an error dialog is displayed.

<img width="477" alt="ReloadError" src="https://user-images.githubusercontent.com/372361/59423280-bf5d0c80-8dc9-11e9-98a1-87f98d863715.png">

Fixes VSTS #588773 - Add support for NuGet SDK resolver

mrward committed:
[Core] Report SDK resolution errors

If an SDK cannot be resolved using the MSBuild SDK resolvers then
an exception is now thrown so the error is reported in the IDE.
Previously the error would be logged in the IDE log and the error
displayed in the IDE would usually be a confusing 'Invalid
configuration mapping'. This also makes the behaviour more consistent
with Visual Studio on Windows.

Fixes VSTS #588773 - Add support for NuGet SDK resolver

Therzok commented on main/src/core/MonoDevelop.Core/MonoDevelop.Projects.MSBuild/SdkResolution.cs:
```diff
@@ -79,17 +80,26 @@ public static SdkResolution GetResolver (TargetRuntime runtime)
 				throw;
 			}
 
+			StringBuilder errorMessage = null;
```
Please use [StringBuilderCache](http://source.monodevelop.com/#MonoDevelop.Core/MonoDevelop.Core/StringBuilderCache.cs,43)

mrward replied:
Changed to use the StringBuilderCache.

sgmunn commented:
@mrward a suggestion here, if the dialog is still shown (master might have disabled showing that anyway) is to derive the exception from UserException which will show a short header and the details of the error in a nicer formatted way.

mrward commented:
This dialog will be shown since this is an error not a warning about an unsupported project. The change in master was not to show the dialog for unsupported project types, which have a special exception.

I can take a look at using the UserException. We would need to add a short message to the exception, something better than VS Win's: 'The project cannot be opened by the project system...', about not being able to resolve the SDK.

Also it looks like the tooltip in the Solution pad would then not have the full details of the exception, just the short message we added, since UserException.Message does not include the details part of the error.

The full detailed errors is only shown in the dialog which is only shown on loading the solution. If you edit the .csproj then you only get the error in the solution pad. So you will only get the full set of errors on initial load. Not sure if that is a good or bad thing. I guess it depends if the short error message has enough information on its own.

mrward committed:
[Core] Improve sdk resolution error message format for user

Use UserException when throwing an SDK not found error. This allows
a short error message and then a longer error details message to
be reported to the user with better formatting. Note that the
Solution pad will only show the short error message. On loading the
project into the IDE the short and longer error messages are
displayed.

mrward commented:
Changed to use a UserException. Screenshots updated.

mrward commented:
@monojenkins backport release-8.2

mrward commented:
@monojenkins backport release-8.2
